ITensorGaussianMPS/test/gmps.jl

```diff
@@ -2,124 +2,192 @@ using ITensorGaussianMPS
 using ITensors
 using LinearAlgebra
 using Test
-
+#using PyPlot
+#matplotlib.use("QtAgg")
 @testset "Basic" begin
   # Test Givens rotations
   v = randn(6)
   g, r = ITensorGaussianMPS.givens_rotations(v)
   @test g * v ≈ r * [n == 1 ? 1 : 0 for n in 1:length(v)]
 end
 
-@testset "Fermion" begin
-  N = 10
-  Nf = N ÷ 2
-
-  # Hopping
-  t = 1.0
-
-  # Hopping Hamiltonian
-  h = Hermitian(diagm(1 => fill(-t, N - 1), -1 => fill(-t, N - 1)))
-  e, u = eigen(h)
-
-  @test h * u ≈ u * Diagonal(e)
-
-  E = sum(e[1:Nf])
-
-  # Get the Slater determinant
-  Φ = u[:, 1:Nf]
-  @test h * Φ ≈ Φ * Diagonal(e[1:Nf])
-
-  # Diagonalize the correlation matrix as a
-  # Gaussian MPS (GMPS)
-  n, gmps = slater_determinant_to_gmps(Φ; maxblocksize=4)
-
-  ns = round.(Int, n)
-  @test sum(ns) == Nf
-
-  Λ = conj(Φ) * transpose(Φ)
-  @test gmps * Λ * gmps' ≈ Diagonal(ns) rtol = 1e-2
-  @test gmps' * Diagonal(ns) * gmps ≈ Λ rtol = 1e-2
-
-  # Form the MPS
-  s = siteinds("Fermion", N; conserve_qns=true)
-  ψ = slater_determinant_to_mps(s, Φ; blocksize=4)
-
+@testset "Hamiltonians" begin
+  # Test Givens rotations
+  N = 8
+  t = -0.8 ###nearest neighbor hopping
+  mu = 0.0 ###on-site chemical potential
+  pairing = 1.2
   os = OpSum()
-  for i in 1:N, j in 1:N
-    if h[i, j] ≠ 0
-      os .+= h[i, j], "Cdag", i, "C", j
+  for i in 1:N
+    if 1 < i < N
+      js = [i - 1, i + 1]
+    elseif i == 1
+      js = [i + 1]
+    else
+      js = [i - 1]
+    end
+    for j in js
+      os .+= t, "Cdag", i, "C", j
+    end
+  end
+  h_hop = ITensorGaussianMPS.hopping_hamiltonian(os)
+  for i in 1:N
+    if 1 < i < N
+      js = [i - 1, i + 1]
+    elseif i == 1
+      js = [i + 1]
+    else
+      js = [i - 1]
+    end
+    for j in js
+      #os .+= pairing, "Cdag", i, "Cdag", j
+      os .+= -conj(pairing), "C", i, "C", j
```

Comment:
Why don't these both have to be specified?

Reply:
I think that for both hopping and pairing terms we should take the inputs very literally and not add in any terms that weren't specified in the `OpSum`. So if someone inputs just `"Cdag", 1, "Cdag" 2` we shouldn't infer that they meant to also add `"C", 1, "C", 2`, and if someone specifies just `"Cdag", 1, "C", 2` we shouldn't infer that they meant to also add `"C", 1, "Cdag", 2`. That is how it works for MPO construction and I think we should be consistent with that (for example, so that a user can input that same `OpSum` into `quadratic_hamiltonian` and `MPO` and get consistent results), and additionally I think we shouldn't be too presumptuous since someone might actually want a non-Hermitian quadratic operator. We can add a check that the hopping Hamiltonian that is extracted is in fact Hermitian where appropriate, or people can check that themselves.

Reply:
Could you explain what is different about how pairing and hopping terms have to be handled? I'm not following that part of the story.

Reply:
`"Cdag", 1, "C", 2` and `"C", 1, "Cdag", 2` are the same term (up to a minus sign due to commutation relations). It's sufficient to have one of them in the `OpSum`. If it's hermitian conjugate is also present, the resulting contribution to the Hamiltonian is also Hermitian, no matter in which order the term was specified (again assuming that the sign from the commutation relations has been taken into account in the coefficient).
`"Cdag", 1, "Cdag", 2` and `"C", 1, "C", 2` are terms which both need to be specified explicitly in the OpSum in order to have a hermitian operator. The old handling of `quadratic_hamiltonian` did not require this, in analogy to the hopping terms above. However, this lead to a mismatch in the definition of the coefficient between an `OpSum` passed to `MPO` and `quadratic_hamiltonian`, so good that you caught this! I fixed it and will push after fixing a weird bug I am running into.

Reply:
Ok thanks for the explanation, that makes sense.

In principle we _could_ have some concept of terms being equivalent up to Hermitian conjugation, which would require defining what the Hermitian conjugate/dagger of a certain operator is, for example with definitions like`dag(::OpName"C") = OpName"Cdag"()`. Then we could take the `dag` of terms of an `OpSum`, and therefore determine which terms are equivalent up to Hermitian conjugation. Of course we don't have to do that for this PR and that would be more involved, just putting it out there in case it inspires some way of designing the code to make it more generic between hopping and pairing terms.

Reply:
The "bug" turned out to be just a normal variation of the quality of the result with parameters, so I think this is resolved, unless you would like me to add a test that actually test for the equivalence of the operator returned by `quadratic_operator(os::OpSum)` and `MPO(os::OpSum)` (`examples/spinless_fermion_pairing.jl` verifies this, but I didn't make it into a test).

Reply:
Thanks for adding the test, that will be a good reference to make sure we're being consistent about the conventions.